dicoogle/src/main/java/pt/ua/dicoogle/core/mlprovider/PrepareDatastoreTask.java

```diff
@@ -0,0 +1,87 @@
+package pt.ua.dicoogle.core.mlprovider;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import pt.ua.dicoogle.plugins.PluginController;
+import pt.ua.dicoogle.sdk.datastructs.SearchResult;
+import pt.ua.dicoogle.sdk.mlprovider.MLDataset;
+import pt.ua.dicoogle.sdk.mlprovider.MLDicomDataset;
+import pt.ua.dicoogle.sdk.mlprovider.MLImageDataset;
+import pt.ua.dicoogle.server.web.dicom.ROIExtractor;
+import pt.ua.dicoogle.server.web.utils.cache.WSICache;
+
+import java.util.HashMap;
+import java.util.UUID;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+
+/**
+ * This task processes Datastore Requests, and builds the appropriate ML Dataset.
+ */
+public class PrepareDatastoreTask implements Callable<MLDataset> {
+
+    private static final Logger logger = LoggerFactory.getLogger(PrepareDatastoreTask.class);
+    private final DatastoreRequest request;
+    private final PluginController controller;
+    private String dataset;
+
+    private final ROIExtractor roiExtractor;
+
+    private final WSICache wsiCache;
+
+    public PrepareDatastoreTask(PluginController controller, DatastoreRequest request) {
+        this.controller = controller;
+        this.request = request;
+        this.dataset = UUID.randomUUID().toString();
+
+        this.wsiCache = WSICache.getInstance();
+        this.roiExtractor = new ROIExtractor();
+    }
+
+    @Override
+    public MLDataset call() throws Exception {
+
+        if(!request.validate())
+            return null;
+
+        switch (request.getDataType()){
+            case DICOM:
+                return new MLDicomDataset(request.getDimLevel(), request.getUids());
+            case IMAGE: // Not operational
+                throw new UnsupportedOperationException("Datastore requests for image objects is not supported");
```

Comment:
This would be used to store normal JPEG images, for providers that do not support DICOM. This can also be used to create datasets of ROIs, as opposed to full DICOM images. Useful in pathology at least.